File: ch2/clip4/boj7562/Main.java
import java.util.LinkedList;
import java.util.Queue;
import java.util.Scanner;

class Main {
    static int n;
    static int[][] board;
    static int[][] visited;
    static int[] dr = {-2, -1, 1, 2, 2, 1, -1, -2};
    static int[] dc = {1, 2, 2, 1, -1, -2, -2, -1};

    public static void main(String[] args) {
        Scanner sc = new Scanner(System.in);

        int tc = sc.nextInt();
        for (int t = 0; t < tc; t++) {
            n = sc.nextInt();
            board = new int[n][n];
            visited = new int[n][n];
            Point start = new Point(sc.nextInt(), sc.nextInt());
            Point end = new Point(sc.nextInt(), sc.nextInt());
            Queue<Point> q = new LinkedList<>();
            q.add(start);
            visited[start.r][start.c] = 1;
            while (!q.isEmpty()) {
                Point now = q.poll();
                if (now.r == end.r && now.c == end.c) {
                    System.out.println(visited[end.r][end.c] - 1);
                    break;
                }
                for (int i = 0; i < 8; i++) {
                    int nr = now.r + dr[i];
                    int nc = now.c + dc[i];
                    if (nr < 0 || nr >= n || nc < 0 || nc >= n) continue;
                    if (visited[nr][nc] == 0) {
                        visited[nr][nc] = visited[now.r][now.c] + 1;
                        q.add(new Point(nr, nc));
                    }
                }
            }
        }
    }
}

class Point {
    int r, c;

    public Point(int r, int c) {
        this.r = r;
        this.c = c;
    }

}

File: ch2/clip4/boj7562/main.cpp
#include <iostream>
#include <queue>

struct point_t
{
    int y;
    int x;
    point_t operator+(point_t rhs)
    {
        return {y + rhs.y, x + rhs.x};
    }
    bool operator==(point_t rhs)
    {
        return y == rhs.y && x == rhs.x;
    }
};

point_t direction[8] = {{2, 1}, {2, -1}, {1, 2}, {1, -2}, {-1, 2}, {-1, -2}, {-2, 1}, {-2, -1}};

int main()
{
    int T, l;
    int depth[300][300];
    point_t start, goal;
    scanf("%d", &T);

    for(int t = 0; t < T; t++)
    {
        scanf("%d %d %d %d %d", &l, &start.y, &start.x, &goal.y, &goal.x);

        for(int j = 0; j < l; j++)
        {
            for(int i = 0; i < l; i++)
            {
                depth[j][i] = 0;
            }
        }

        std::queue<point_t> point;
        depth[start.y][start.x] = 1;
        
        for(point.push(start); ; point.pop())
        {
            point_t now = point.front();

            if(now == goal)
            {
                printf("%d\n", depth[goal.y][goal.x] - 1);
                break;
            }

            for(int i = 0; i < 8; i++)
            {
                point_t next =  now + direction[i];
                if(next.y >= 0 && next.y < l && next.x >= 0 && next.x < l)
                {
                    if(0 == depth[next.y][next.x])
                    {
                        depth[next.y][next.x] = depth[now.y][now.x] + 1;
                        point.push({next.y, next.x});
                    }
                }
            }
        }        
    }

    return 0;
}